src/common/io/src/buffer/buffer_read_datetime_ext.rs

```diff
@@ -61,48 +70,199 @@ where R: BufferRead
     }
 
     fn read_timestamp_text(&mut self, tz: &Tz) -> Result<DateTime<Tz>> {
-        let mut buf = vec![0; DATE_TIME_LEN];
+        // Date Part YYYY-MM-DD
+        let mut buf = vec![0; DATE_LEN];
         self.read_exact(buf.as_mut_slice())?;
+        let mut v = std::str::from_utf8(buf.as_slice())
+            .map_err_to_code(ErrorCode::BadBytes, || {
+                format!("Cannot convert value:{:?} to utf8", buf)
+            })?;
 
-        let v = std::str::from_utf8(buf.as_slice())
-            .map_err_to_code(ErrorCode::BadBytes, || "Cannot convert value to utf8")?;
-        // convert zero timestamp to `1970-01-01 00:00:00`
-        let res = if v == "0000-00-00 00:00:00" {
-            tz.from_utc_datetime(&NaiveDate::from_ymd(1970, 1, 1).and_hms(0, 0, 0))
-        } else {
-            tz.datetime_from_str(v, "%Y-%m-%d %H:%M:%S")
-                .or_else(|_| tz.datetime_from_str(v, "%Y-%m-%dT%H:%M:%S"))
-                .map_err_to_code(ErrorCode::BadBytes, || {
-                    format!("Cannot parse value:{:?} to DateTime type", v)
-                })?
+        // convert zero date to `1970-01-01`
+        if v == "0000-00-00" {
+            v = "1970-01-01";
+        }
+        let d = v
+            .parse::<NaiveDate>()
+            .map_err_to_code(ErrorCode::BadBytes, || {
+                format!("Cannot parse value:{} to Date type", v)
+            })?;
+        let mut dt = tz.from_local_datetime(&d.and_hms(0, 0, 0)).unwrap();
+
+        let less_1000 = |dt: DateTime<Tz>| {
+            // convert timestamp less than `1000-01-01 00:00:00` to `1000-01-01 00:00:00`
+            if dt.year() < 1000 {
+                Ok(tz.from_utc_datetime(&NaiveDate::from_ymd(1000, 1, 1).and_hms(0, 0, 0)))
+            } else {
+                Ok(dt)
+            }
         };
 
-        self.ignore(|b| b == b'z' || b == b'Z')?;
-        let dt = if self.ignore_byte(b'.')? {
-            buf.clear();
+        // Time Part
+        buf.clear();
+        if self.ignore(|b| b == b' ' || b == b'T').unwrap() {
+            // HH:mm:ss
+            let get_time = |buf: &mut Vec<u8>, size| {
+                if size > 0 && size < 3 {
+                    Ok(lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap())
+                } else {
+                    Err(ErrorCode::BadBytes(
+                        "err with parse time part. Format like this:[03:00:00]",
+                    ))
+                }
+            };
+            let mut buf = Vec::with_capacity(2);
             let size = self.keep_read(&mut buf, |f| (b'0'..=b'9').contains(&f))?;
-            let scales: i64 = lexical_core::FromLexical::from_lexical(buf.as_slice()).unwrap();
+            match get_time(&mut buf, size) {
```

Comment:
why not use `get_time(&mut buf, size).and_then ....`

This part of code looks very ugly.

Reply:
Yes, and_then can simplify match arm. 

BTW, could we have other ideas to simplify this?

```
match res {
    Ok(d) => dt = d,
    Err(e) => return Err(e),
}
```